woocommerce/woocommerce
# [Experimental] Reliability fixes for new filter blocks

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR is extracted from #51624 while working on #51622. This PR includes various bug fixes to improve the stability and reliability of new filter blocks powered by interactivity API. I break the original PR down into two separate PRs to make them easier to review.

This PR:

- adds the `getInnerBlockBy` function enabling retrieving inner blocks by callback.
- updates the naming convention for the target attribute key when querying element value.
- removes unused `clear-button/frontend.ts` file.
- updates block context `filterData` type.
- adds the `data-wc-key` attribute to each filter block to prevent the diffing issue.
- fixes the styling issues after navigation by always rendering all inner blocks (even the block doesn't have any filter options) and hiding the block using the `hidden` attribute.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

0. Ensure the experimental blocks are enabled:
    - If you use the WooCommerce Beta Tester plugin, you can enable the experimental blocks in Tools > WCA Test Helper > Features. Toggle the `experimental-blocks` option.
    - If you build this PR locally, build the branch with the `pnpm --filter='@woocommerce/plugin-woocommerce' watch:build` command to have experimental blocks available.
1.  Ensure the `Product Filters` template part is reset.
2. Add another Attribute filter block to the template part, and select a different attribute.
3. Inside that new attribute filter block, add a block that doesn't exist on the page, for example, Cover block.
4. Add the `Product Filters (Experimental)` template part to the `Product Catalog` template.
5. Save and go to the shop page.
6. Select an attribute, see the URL updated and a clear button appears.
7. Click on the clear button clears the selected filters.
8. On the first attribute filter block that doesn't contain the cover block, select an attribute that make the second attribute filter block disappear.
9. Reload the page, see the selected filter persists and the second attribute filter block is still hidden.
10. Clear the selected filter, and see the second filter block appear with the cover block style is loaded correctly.



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
